platform/osx/os_osx.mm

```diff
@@ -610,96 +611,101 @@ static int translateKey(unsigned int key) {
 		for (i = 0; i < length; i++) {
-			ev.key.unicode = [characters characterAtIndex:i];
-			OS_OSX::singleton->push_input(ev);
-			ev->get_scancode() = 0;
+			k->set_unicode([characters characterAtIndex:i]);
+			OS_OSX::singleton->push_input(k);
+			k->set_scancode(0);
 		}
 	} else {
-		OS_OSX::singleton->push_input(ev);
+		OS_OSX::singleton->push_input(k);
 	}
 }
 
 - (void)flagsChanged:(NSEvent *)event {
-	Ref<InputEvent> ev;
+	Ref<InputEventKey> k;
+	k.instance();
+	
 	int key = [event keyCode];
 	int mod = [event modifierFlags];
 
-	ev.type = InputEvent::KEY;
-
 	if (key == 0x36 || key == 0x37) {
 		if (mod & NSCommandKeyMask) {
 			mod &= ~NSCommandKeyMask;
-			ev->is_pressed() = true;
+			k->set_pressed(true);
 		} else {
-			ev->is_pressed() = false;
+			k->set_pressed(false);
 		}
 	} else if (key == 0x38 || key == 0x3c) {
 		if (mod & NSShiftKeyMask) {
 			mod &= ~NSShiftKeyMask;
-			ev->is_pressed() = true;
+			k->set_pressed(true);
 		} else {
-			ev->is_pressed() = false;
+			k->set_pressed(false);
 		}
 	} else if (key == 0x3a || key == 0x3d) {
 		if (mod & NSAlternateKeyMask) {
 			mod &= ~NSAlternateKeyMask;
-			ev->is_pressed() = true;
+			k->set_pressed(true);
 		} else {
-			ev->is_pressed() = false;
+			k->set_pressed(false);
 		}
 	} else if (key == 0x3b || key == 0x3e) {
 		if (mod & NSControlKeyMask) {
 			mod &= ~NSControlKeyMask;
-			ev->is_pressed() = true;
+			k->set_pressed(true);
 		} else {
-			ev->is_pressed() = false;
+			k->set_pressed(false);
 		}
 	} else {
 		return;
 	}
 
-	ev.key.mod = translateFlags(mod);
-	ev->get_scancode() = latin_keyboard_keycode_convert(translateKey(key));
+	get_osx_key_modifier_state(mod, k);
+	k->set_scancode(latin_keyboard_keycode_convert(translateKey(key)));
 
-	OS_OSX::singleton->push_input(ev);
+	OS_OSX::singleton->push_input(k);
 }
 
 - (void)keyUp:(NSEvent *)event {
 
-	Ref<InputEvent> ev;
-	ev.type = InputEvent::KEY;
-	ev->is_pressed() = false;
-	ev.key.mod = translateFlags([event modifierFlags]);
-	ev->get_scancode() = latin_keyboard_keycode_convert(translateKey([event keyCode]));
-	OS_OSX::singleton->push_input(ev);
+	Ref<InputEventKey> k;
+	k.instance();
+
+	get_osx_key_modifier_state([event modifierFlags], k);
+	k->set_pressed(false);
+	k->set_scancode(latin_keyboard_keycode_convert(translateKey([event keyCode])));
+
+	OS_OSX::singleton->push_input(k);
 }
 
 inline void sendScrollEvent(int button, double factor) {
-	Ref<InputEvent> ev;
-	ev.type = Ref<InputEvent>::MOUSE_BUTTON;
-	ev->get_button_index() = button;
-	ev.mouse_button->get_factor() = factor;
-	ev->is_pressed() = true;
-	ev->get_pos().x = mouse_x;
-	ev->get_pos().y = mouse_y;
-	ev.mouse_button.global_x = mouse_x;
-	ev.mouse_button.global_y = mouse_y;
-	ev->get_button_mask() = button_mask;
-	OS_OSX::singleton->push_input(ev);
-	ev->is_pressed() = false;
-	OS_OSX::singleton->push_input(ev);
+	//toger scroll event has not modifier keys or are they added somewhere else
```

Comment:
If there are any, I guess you should retrieve them in `(void)scrollWheel:(NSEvent *)event` as done for other events, and pass them as argument to the helper `sendScrollEvent`.